### PR TITLE
feat: parallel coordinates chart on trial hyperparameters tab [WEB-1196]

### PR DESCRIPTION
**Pull request by EmilyBonar**
## Description
Task
Update the viz on the single trial “Hyperparameters” tab for single trials that are part of a multi-trial search from current vis to pre-filtered parallel coordinates chart Show PC chart.

Result
Screenshot 2023-05-19 at 15-24-24 Trial 5820 - Determined

## Test Plan

- Navigate to a Trial Details page inside a multi-trial experiment
- Go to Hyperparameters tab
- It should look like the screenshot above, with the current trial line being active and all others faded but visible
- Chart should **not** be interactable (ie reordering dimensions and creating filters), but all selects in chart header should work as normal
- Below chart there should be a table showing exact values of current trial hyperparameters (the list of hps might differ from the chart because constant hyperparameters are not shown in parallel coordinates)

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

**Comment by alexjdetermined** [comment]
Issues
1. **Some trials not plotting:** Not plotting for some trials and unsure why (the hp table shows values for the layers but line not showing up): [example trial that wont plot](https://deploy-preview-6872--determined-ui.netlify.app/experiments/139/trials/373/hyperparameters)
2. **Chart loads twice/flicker:** Noticing that when flipping between trials, the chart seems to load then flicker/reload again one to a few times. 
3. Tab switch causes error: Go to a trial, go to the HP tab then switch to any other tab in the trial (e.g. overview). Result is getting a Hermes related unexpected application error. Unclear if netlify issue or not.

File: webui/react/src/pages/TrialDetails/MultiTrialDetailsHyperparameters.settings.ts [hunk 1]
```diff
@@ -1,50 +1,61 @@
...
+      type: union([
+        undefinedType,
+        type({ name: string, type: keyof({ training: null, validation: null }) }),
+      ]),
+    },
+    scale: {
+      defaultValue: Scale.Linear,
+      storageKey: 'scale',
+      type: keyof({ linear: null, log: null }),
+      // See https://github.com/gcanti/io-ts/blob/master/index.md#union-of-string-literals
```
Comment (hkang1) [inline webui/react/src/pages/TrialDetails/MultiTrialDetailsHyperparameters.settings.ts:58]: 👍

File: webui/react/src/pages/TrialDetails/MultiTrialDetailsHyperparameters.settings.ts [hunk 2]
```diff
@@ -1,50 +1,61 @@
...
 const DEFAULT_MAX_TRIALS = 100;
 
 export interface ExperimentHyperparametersSettings {
-  filters: VisualizationFilters;
+  batch: number;
+  batchMargin: number;
+  hParams: string[];
+  maxTrial: number;
+  metric?: Metric;
+  scale: Scale;
```
Comment (hkang1) [inline webui/react/src/pages/TrialDetails/MultiTrialDetailsHyperparameters.settings.ts:16]: nice I was going to suggest flattening this in the previous commit but you already did it here

File: webui/react/src/pages/TrialDetails/MultiTrialDetailsHyperparameters.tsx [hunk 3]
```diff
@@ -95,13 +116,13 @@ const MultiTrialDetailsHyperparameters: React.FC<Props> = ({
...
-  }, [experiment.id, isSupported, settings.filters.metric]);
+  }, [experiment.id, isSupported, settings.metric]);
 
   // Set the default filter batch.
   useEffect(() => {
-    if (!batches || batches.length === 0 || settings.filters.batch !== 0) return;
-    updateSettings({ filters: { ...settings.filters, batch: batches.first() } });
-  }, [batches, settings.filters, updateSettings]);
+    if (!batches || batches.length === 0 || settings.batch !== 0) return;
+    updateSettings({ batch: batches.first() });
```
Comment (hkang1) [inline webui/react/src/pages/TrialDetails/MultiTrialDetailsHyperparameters.tsx:123]: **question:** do you know if `updateSettings` takes a full setting or partial deltas? I'm looking at the code and I think it's implemented incorrectly? I looks like it can take in partial but the check of whether it's changed or not is assuming it's a full settings object...
```
  const updateSettings = useCallback(
    (updates: Settings) => {
      stateOb.update((s) =>
        s.set(
          config.storagePath,
          isEqual(s.get(config.storagePath), updates)
            ? s.get(config.storagePath) ?? {}
            : { ...s.get(config.storagePath), ...updates },
        ),
      );
    },
    [config, stateOb],
  );
```

File: webui/react/src/pages/TrialDetails/MultiTrialDetailsHyperparameters.tsx [hunk 4]
```diff
@@ -57,15 +56,15 @@ const MultiTrialDetailsHyperparameters: React.FC<Props> = ({
...
+  const filters: VisualizationFilters = useMemo(
     () => ({
       batch: settings.batch,
       batchMargin: settings.batchMargin,
       hParams: settings.hParams,
       maxTrial: settings.maxTrial,
       metric: settings.metric,
       scale: settings.scale,
-      view: ViewType.Grid,
+      view: ViewType.Grid, // View is required in the type but not used in parallel coordinates
```
Comment (hkang1) [inline webui/react/src/pages/TrialDetails/MultiTrialDetailsHyperparameters.tsx:67]: maybe this could be made optional and default to `ViewType.Grid`

**Review by hkang1 — COMMENTED** [review]
I don't have anything block from code perspective other than some minor things. Holding off a stamp just for now, for the design issues.

**Review by hkang1 — COMMENTED** [review]
Looks great! I don't have anything blocking from code perspective other than some minor things. Holding off a stamp just for now, for the design issues.

File: webui/react/src/pages/TrialDetails/MultiTrialDetailsHyperparameters.tsx (hunk 3 above)
Comment (EmilyBonar) [reply webui/react/src/pages/TrialDetails/MultiTrialDetailsHyperparameters.tsx:123]: My understanding is that it can take a delta, and from a cursory reading of that code I think that passing in a delta should work. At worst it seems like there might be a slight inefficiency in the case where you pass in a delta where the key/value pair matches the current state.

File: webui/react/src/pages/ExperimentDetails/ExperimentVisualization/HpScatterPlots.tsx [hunk 5]
```diff
@@ -28,15 +28,15 @@ interface Props {
...
+  selectedMetric?: Metric;
   selectedScale: Scale;
 }
 
 interface HpMetricData {
   hpLabels: Record<string, string[]>;
   hpLogScales: Record<string, boolean>;
   hpValues: Record<string, number[]>;
-  metricValues: Record<string, (number | null)[]>;
+  metricValues: Record<string, (number | undefined)[]>;
```
[the remark this answers is not in the document]
Comment (EmilyBonar) [reply webui/react/src/pages/ExperimentDetails/ExperimentVisualization/HpScatterPlots.tsx:39]: Yes, according to Swagger the API can only return `number`s. The null values were introduced in this file on line 175 to filter out NaN/Infinities. All I did here was switch from using null to undefined.

File: webui/react/src/pages/TrialDetails/MultiTrialDetailsHyperparameters.tsx [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+
+  const [batches, setBatches] = useState<number[]>();
+
+  // Stream available metrics.
+  const metrics = useMetricNames(experiment.id, handleError);
+
+  const isSupported = useMemo(() => {
+    return !(
+      ExperimentSearcherName.Single === experiment.config.searcher.name ||
+      ExperimentSearcherName.Pbt === experiment.config.searcher.name
```
[the remark this answers is not in the document]
Comment (hkang1) [reply webui/react/src/pages/TrialDetails/MultiTrialDetailsHyperparameters.tsx:83]: I guess it doesn't hurt to just leave this in. 🤷

File: webui/react/src/pages/TrialDetails/MultiTrialDetailsHyperparameters.tsx (hunk 3 above)
Comment (hkang1) [reply webui/react/src/pages/TrialDetails/MultiTrialDetailsHyperparameters.tsx:121]: FYI, [related PR](https://github.com/determined-ai/determined/pull/6894) that stemmed from this conversation :D

TL;DR - code-wise it was only supporting full settings but due to incorrect typings it allowed partials through.

**Review by hkang1 — APPROVED** [review]
Code looks good!

I was able to repro what Alex brought up, especially with the target trial not showing up for some trials. Can you take a look at that too please.

**Review by alexjdetermined — APPROVED** [review]
LMTM! nice work around for the batch display issue.

File: webui/react/src/pages/ExperimentDetails/ExperimentVisualization/ExperimentVisualizationFilters.tsx [hunk 7]
```diff
@@ -139,84 +89,88 @@ const ExperimentVisualizationFilters: React.FC<Props> = ({
...
+  const handleScaleChange = useCallback(
+    (scale: Scale) => {
+      onChange?.({ scale });
+    },
+    [onChange],
+  );
 
   const handleReset = useCallback(() => {
-    dispatch({ type: ActionType.Set, value: filters });
+    //dispatch({ type: ActionType.Set, value: filters });
```
Comment (hkang1) [inline webui/react/src/pages/ExperimentDetails/ExperimentVisualization/ExperimentVisualizationFilters.tsx:146]: **nit:** delete

File: webui/react/src/pages/ExperimentDetails/ExperimentVisualization/ExperimentVisualizationFilters.tsx [hunk 8]
```diff
@@ -139,84 +89,88 @@ const ExperimentVisualizationFilters: React.FC<Props> = ({
...
-  }, [filters, onReset]);
+  }, [onReset]);
 
   // Pick the first valid option if the current local batch is invalid.
   useEffect(() => {
-    if (batches.length === 0 || batches.includes(localFilters.batch)) return;
-    dispatch({ type: ActionType.SetBatch, value: batches.first() });
-  }, [batches, localFilters.batch]);
+    if (batches.length === 0 || batches.includes(filters.batch)) return;
+    //dispatch({ type: ActionType.SetBatch, value: batches.first() });
```
Comment (hkang1) [inline webui/react/src/pages/ExperimentDetails/ExperimentVisualization/ExperimentVisualizationFilters.tsx:153]: ditto

File: webui/react/src/components/ParallelCoordinates.tsx [hunk 9]
```diff
@@ -26,15 +26,17 @@ const ParallelCoordinates: React.FC<Props> = ({
...
     return () => {
       chartRef.current?.destroy();
       chartRef.current = undefined;
     };
-  }, [dimensions, disableInteraction]);
+  }, []);
+
+  useEffect(() => {
+    if (disableInteraction) chartRef.current?.disable();
+  }, [disableInteraction]);
```
Comment (hkang1) [inline webui/react/src/components/ParallelCoordinates.tsx:39]: this works ok now? Didn't you have issues with this approach originally?

**Review by hkang1 — CHANGES_REQUESTED** [review]
The changes, code-wise, look good and rendering looks great, but something is not behaving properly on the chart side. So as an example, if you navigate to[this preview link](https://deploy-preview-6872--determined-ui.netlify.app/experiments/139/visualization/hp-parallel-coordinates), try to play around with the filters on the pc chart. If you compare it with our current version, the on change filter events are not registering. Previously, when you create a filter, you can dynamically see it changing as you click and drag but now the filter doesn't show up until you release the drag click. Same thing with the filter move (drag and drop). Maybe the filterChange event isn't being handled?

File: webui/react/src/components/ParallelCoordinates.tsx (hunk 9 above)
Comment (EmilyBonar) [reply webui/react/src/components/ParallelCoordinates.tsx:39]: I did at first, but after the refactor it shouldn't be a problem because the chart isn't getting destroyed and recreated. I'm not seeing any issue wrt this in testing, but if you see something let me know.

**Comment by hkang1** [comment]
> The changes, code-wise, look good and rendering looks great, but something is not behaving properly on the chart side. So as an example, if you navigate to[this preview link](https://deploy-preview-6872--determined-ui.netlify.app/experiments/139/visualization/hp-parallel-coordinates), try to play around with the filters on the pc chart. If you compare it with our current version, the on change filter events are not registering. Previously, when you create a filter, you can dynamically see it changing as you click and drag but now the filter doesn't show up until you release the drag click. Same thing with the filter move (drag and drop). Maybe the filterChange event isn't being handled?

huh I don't know what it was, but after a full reboot I can' repro it...

**Review by hkang1 — APPROVED** [review]
The changes look good and works great. Came across some other artifacts but they are not unique to this PR, but will not them just to have a reference for future work.

- Non-scalar HP values (such as JSON values) will render very long labels [example](https://deploy-preview-6872--determined-ui.netlify.app/experiments/133/visualization/hp-parallel-coordinates). They seem to also cause other valid metrics to not plot properly despite filtering out the non-scalar HP columns.
- Batches Processed and Batch Margin filters appear on the trial HP page, but it might not make sense to make them available. This would be more of a product question, but maybe they are better to have for reference so the user is at least aware of the batch ranges the trial was generated from so they can better compare apples to apples.

**Comment by hkang1** [comment]
Can confirm the creation and destruction of Hermes is well done!
